rp2040-hal/src/i2c/peripheral.rs

```diff
@@ -0,0 +1,199 @@
+        i2c.ic_con.modify(|_, w| {
+            // run in fast mode
+            w.speed().fast();
+            // setup slave mode
+            w.master_mode().disabled();
+            w.ic_slave_disable().slave_enabled();
+            // hold scl when fifo's full
+            w.rx_fifo_full_hld_ctrl().enabled();
+            w.ic_restart_en().enabled();
+            w.tx_empty_ctrl().enabled()
+        });
+
+        // Clear FIFO threshold
+        i2c.ic_tx_tl.write(|w| unsafe { w.tx_tl().bits(0) });
+        i2c.ic_rx_tl.write(|w| unsafe { w.rx_tl().bits(0) });
+
+        // Enable IP
+        i2c.ic_enable.write(|w| w.enable().enabled());
+
+        I2CPeripheralEventIterator {
+            i2c: Self {
+                i2c,
+                pins: (sda_pin, scl_pin),
+                mode: PhantomData,
+            },
+            state: State::Idle,
+        }
+    }
+}
+
+impl<T: Deref<Target = I2CBlock>, PINS> I2CPeripheralEventIterator<T, PINS> {
+    /// Pushs up to `usize::min(TX_FIFO_SIZE, buf.len())` bytes to the TX FIFO.
+    /// Returns the number of bytes pushed to the FIFO. Note this does *not* reflect how many bytes
+    /// are effectively received by the controller.
+    pub fn write(&mut self, buf: &[u8]) -> usize {
+        // just in case, clears previous tx abort.
+        self.i2c.i2c.ic_clr_tx_abrt.read();
+
+        let mut sent = 0;
+        for &b in buf.iter() {
+            if self.i2c.tx_fifo_full() {
+                break;
+            }
+
+            self.i2c
+                .i2c
+                .ic_data_cmd
+                .write(|w| unsafe { w.dat().bits(b) });
+            sent += 1;
+        }
+        // serve a pending read request
+        self.i2c.i2c.ic_clr_rd_req.read();
+        sent
+    }
+
+    /// Pulls up to `usize::min(RX_FIFO_SIZE, buf.len())` bytes from the RX FIFO.
+    pub fn read(&mut self, buf: &mut [u8]) -> usize {
+        let mut read = 0;
+
+        for b in buf.iter_mut() {
+            if self.i2c.rx_fifo_empty() {
+                break;
+            }
+
+            *b = self.i2c.i2c.ic_data_cmd.read().dat().bits();
+            read += 1;
+        }
+        read
+    }
+}
+impl<T: Deref<Target = I2CBlock>, PINS> Iterator for I2CPeripheralEventIterator<T, PINS> {
+    type Item = I2CEvent;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let stat = self.i2c.i2c.ic_raw_intr_stat.read();
+        self.i2c.i2c.ic_clr_activity.read();
+
+        match self.state {
+            State::Idle if stat.start_det().bit_is_set() => {
+                self.i2c.i2c.ic_clr_start_det.read();
+                self.state = if stat.rd_req().bit_is_set() {
+                    State::Read
+                } else {
+                    State::Write
+                };
+                Some(I2CEvent::Start)
+            }
+            State::Read if stat.rd_req().bit_is_set() => {
+                // Bit is cleared by a call to write
+                // self.0.i2c.ic_clr_rd_req.read();
```

Comment:
No this is due to how the hardware is made. On a read request, clearing this bit indicates the I2C Block that data has been pushed to the FIFO and it can stop clock-stretching.
I'll update the comment to make that more obvious.

Reply:
Updated, let me know what you think about this new version.

Reply:
Yep much clearer, thanks!